app/assets/javascripts/school_administrator_dashboard/discipline_dashboard/SchoolDisciplineDashboard.js

```diff
@@ -41,12 +45,56 @@ export default class SchoolDisciplineDashboard extends React.Component {
     this.setState({selectedChart: selection.value, selectedCategory: null});
   }
 
-  filterIncidentDates(incidentsArray) {
-    const {nowFn} = this.context;
-    const {timeRangeKey} = this.state;
-    const range = momentRange(timeRangeKey, nowFn());
-    return incidentsArray.filter((incident) => {
-      return moment.utc(incident.occurred_at).isBetween(range[0], range[1]);
+  allDisciplineIncidents() {
+    return this.memoize(['allDisciplineIncidents'], () => {
+      const {dashboardStudents} = this.props;
+      return _.flattenDeep(_.compact(dashboardStudents.map(student => this.mergeDisciplineData(student.discipline_incidents, student))));
```

Comment:
Can you say more about why the transformation in `allDisciplineIncidents` and `mergeDisciplineData` is necessary to do up front?

It looks to me like the flattening has to get undone later one with regrouping by student, and the merge fn looks to me like it's changing the names of the keys and doing little transformations like rendering a moment as a user-facing string.  These seem like they could be moved from being precomputed up front to done on-demand wherever they are needed in the `render` methods.

Reply:
Sure thing! The problem I'm trying to solve is that the groups we want aren't attributes of a single object. Some are student attributes , some are discipline incident attributes, and some are computed from the discipline incident. I find it more readable and less confusing to have one function that "owns" all the attributes we care about for grouping and then just _groupBy to get our categories. That way whenever we want to add some other category, we have a place to put that logic and see it all together.

This isn't exactly precomputed since this method is only called in the main render method and the results are passed down to the other methods that need them. Unless I'm misunderstanding something it should still be happening on demand (that's why I removed the old SchoolWideDisciplineIncidents component that really was precomputing all this stuff).